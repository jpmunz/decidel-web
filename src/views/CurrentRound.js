import React from "react";
import { connect } from "react-redux";
import { Button, Alert } from "react-bootstrap";
import LoadingButton from "./LoadingButton";
import "./CurrentRound.scss";
import autoBind from "react-autobind";
import {
  setCurrentUserNameRequest,
  getAsyncRequest,
  persistCurrentRoundRequest,
  isPersistable,
  getFinalDecision,
  isCurrentTurn,
  getCurrentDeciderName
} from "src/state";

export class CurrentRound extends React.Component {
  constructor(props) {
    super(props);
    autoBind(this);
  }

  handleSetCurrentUserClick() {
    this.props.setCurrentUserNameRequest({
      name: this.props.currentDeciderName
    });
  }

  handlePersistClick() {
    this.props.persistCurrentRoundRequest({
      id: this.props.id
    });
  }

  render() {
    let content;

    if (this.props.finalDecision) {
      content = (
        <>
          <span>{"You decided on " + this.props.finalDecision}</span>
        </>
      );
    } else if (this.props.isCurrentTurn) {
      content = (
        <>
          <span>
            {this.props.request.inProgress
              ? ""
              : "It's your turn - narrow down that decision!"}
          </span>
          <LoadingButton
            disabled={!this.props.isPersistable}
            title={
              this.props.isPersistable
                ? ""
                : "You need to do at least one thing on your turn"
            }
            loading={this.props.request.inProgress}
            onClick={this.handlePersistClick}
            text="I'm Done"
          />
        </>
      );
    } else {
      content = (
        <>
          <span>It's </span>
          <span className="user-name">{this.props.currentDeciderName}</span>
          <span>'s turn - wait for them to cross something off.</span>
          <Button onClick={this.handleSetCurrentUserClick}>That's Me!</Button>
        </>
      );
    }

    let alert;
    if (this.props.request.error) {
      alert = (
        <Alert variant="warning">
          There was a problem saving your changes. Please try again later.
        </Alert>
      );
    }

    return (
      <div className="CurrentRound">
        {alert}
        {content}
      </div>
    );
  }
}

export default connect(
  (state, props) => {
    return {
      finalDecision: getFinalDecision(state, props.id),
      request: getAsyncRequest(state, props.id),
      isCurrentTurn: isCurrentTurn(state, props.id),
      isPersistable: isPersistable(state, props.id),
      currentDeciderName: getCurrentDeciderName(state, props.id)
    };
  },
  { persistCurrentRoundRequest, setCurrentUserNameRequest }
)(CurrentRound);
