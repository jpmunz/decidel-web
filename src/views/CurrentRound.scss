@import "mixins";

.CurrentRound {
  text-align: right;
  margin-bottom: 10px;

  .user-name {
    font-weight: bold;
  }

  .btn {
    margin-left: 10px;
    @include box-shadow(5px, 5px, 10px, rgba(0, 0, 0, 0.2));
  }
}
