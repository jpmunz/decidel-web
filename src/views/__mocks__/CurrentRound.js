import React from "react";

export const MockCurrentRound = jest.fn();

MockCurrentRound.mockReturnValue(<div>Mocked CurrentRound</div>);

const mock = jest.fn().mockImplementation(MockCurrentRound);

export default mock;
