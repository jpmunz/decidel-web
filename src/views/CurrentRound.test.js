import React from "react";
import { CurrentRound } from "./CurrentRound";
import { render, fireEvent } from "@testing-library/react";

const persistMock = jest.fn();
const setUserNameMock = jest.fn();

const renderComponent = function({
  id = "22",
  finalDecision = false,
  isPersistable = true,
  isCurrentTurn = true,
  request = {},
  currentDeciderName = "Bob"
}) {
  return render(
    <CurrentRound
      id={id}
      finalDecision={finalDecision}
      isPersistable={isPersistable}
      isCurrentTurn={isCurrentTurn}
      request={request}
      currentDeciderName={currentDeciderName}
      persistCurrentRoundRequest={persistMock}
      setCurrentUserNameRequest={setUserNameMock}
    />
  );
};

const renderTest = options =>
  expect(renderComponent(options).asFragment()).toMatchSnapshot();

test("Current user's turn but current round is not persistable", () =>
  renderTest({ isCurrentTurn: true, isPersistable: false }));

test("Current round persisting in progress", () =>
  renderTest({ request: { inProgress: true } }));

test("Current round failed to persist", () =>
  renderTest({ request: { error: "failed" } }));

test("Another user's turn", () => renderTest({ isCurrentTurn: false }));

test("Final decision has been made", () =>
  renderTest({ finalDecision: "The Rock" }));

test("Trigger persisting the current round", () => {
  const component = renderComponent({
    id: "22",
    isPersistable: true,
    isCurrentTurn: true
  });

  expect(component.asFragment()).toMatchSnapshot();

  fireEvent.click(component.getByText("I'm Done"));
  expect(persistMock.mock.calls).toEqual([[{ id: "22" }]]);
});

test("Setting the current user", () => {
  const component = renderComponent({
    isCurrentTurn: false,
    currentDeciderName: "Phil"
  });
  fireEvent.click(component.getByText("That's Me!"));
  expect(setUserNameMock.mock.calls).toEqual([[{ name: "Phil" }]]);
});
