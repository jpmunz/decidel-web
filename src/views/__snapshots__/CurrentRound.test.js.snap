// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Another user's turn 1`] = `
<DocumentFragment>
  <div
    class="CurrentRound"
  >
    <span>
      It's 
    </span>
    <span
      class="user-name"
    >
      Bob
    </span>
    <span>
      's turn - wait for them to cross something off.
    </span>
    <button
      class="btn btn-primary"
      type="button"
    >
      That's Me!
    </button>
  </div>
</DocumentFragment>
`;

exports[`Current round failed to persist 1`] = `
<DocumentFragment>
  <div
    class="CurrentRound"
  >
    <div
      class="fade alert alert-warning show"
      role="alert"
    >
      There was a problem saving your changes. Please try again later.
    </div>
    <span>
      It's your turn - narrow down that decision!
    </span>
    <button
      class="btn btn-primary"
      title=""
      type="button"
    >
      I'm Done
    </button>
  </div>
</DocumentFragment>
`;

exports[`Current round persisting in progress 1`] = `
<DocumentFragment>
  <div
    class="CurrentRound"
  >
    <span />
    <button
      class="btn btn-primary"
      title=""
      type="button"
    >
      <span
        aria-hidden="true"
        class="spinner-border spinner-border-sm"
        role="status"
      />
      <span
        class="sr-only"
      >
        Loading...
      </span>
    </button>
  </div>
</DocumentFragment>
`;

exports[`Current user's turn but current round is not persistable 1`] = `
<DocumentFragment>
  <div
    class="CurrentRound"
  >
    <span>
      It's your turn - narrow down that decision!
    </span>
    <button
      class="btn btn-primary"
      disabled=""
      title="You need to do at least one thing on your turn"
      type="button"
    >
      I'm Done
    </button>
  </div>
</DocumentFragment>
`;

exports[`Final decision has been made 1`] = `
<DocumentFragment>
  <div
    class="CurrentRound"
  >
    <span>
      You decided on The Rock
    </span>
  </div>
</DocumentFragment>
`;

exports[`Trigger persisting the current round 1`] = `
<DocumentFragment>
  <div
    class="CurrentRound"
  >
    <span>
      It's your turn - narrow down that decision!
    </span>
    <button
      class="btn btn-primary"
      title=""
      type="button"
    >
      I'm Done
    </button>
  </div>
</DocumentFragment>
`;
